x/crosschain/keeper/abci.go

```diff
@@ -28,38 +42,59 @@ func (k Keeper) IterateAndUpdateCctxGasPrice(ctx sdk.Context) error {
 
 	// skip if haven't reached epoch end
 	if ctx.BlockHeight()%gasPriceIncreaseFlags.EpochLength != 0 {
-		return nil
+		return 0, gasPriceIncreaseFlags
 	}
 
-	// iterate all chains' pending cctx
-	chains := common.DefaultChainsList()
+	cctxCount := 0
+
+IterateChains:
 	for _, chain := range chains {
-		res, err := k.CctxListPending(sdk.UnwrapSDKContext(ctx), &types.QueryListCctxPendingRequest{
-			ChainId: chain.ChainId,
-			Limit:   gasPriceIncreaseFlags.MaxPendingCctxs,
-		})
-		if err != nil {
-			return err
-		}
+		// support only external evm chains
+		if common.IsEVMChain(chain.ChainId) && !common.IsZetaChain(chain.ChainId) {
+			res, err := k.CctxListPending(sdk.UnwrapSDKContext(ctx), &types.QueryListCctxPendingRequest{
+				ChainId: chain.ChainId,
+				Limit:   gasPriceIncreaseFlags.MaxPendingCctxs,
+			})
+			if err != nil {
+				ctx.Logger().Info("GasStabilityPool: fetching pending cctx failed",
+					"chainID", chain.ChainId,
+					"err", err.Error(),
+				)
+				continue IterateChains
+			}
 
-		// iterate through all pending cctx
-		for _, pendingCctx := range res.CrossChainTx {
-			if pendingCctx != nil {
-				_, _, err := k.CheckAndUpdateCctxGasPrice(ctx, *pendingCctx, gasPriceIncreaseFlags)
-				if err != nil {
-					return err
+			// iterate through all pending cctx
+			for _, pendingCctx := range res.CrossChainTx {
+				if pendingCctx != nil {
+					gasPriceIncrease, additionalFees, err := updateFunc(ctx, k, *pendingCctx, gasPriceIncreaseFlags)
+					if err != nil {
+						ctx.Logger().Info("GasStabilityPool: updating gas price for pending cctx failed",
+							"cctxIndex", pendingCctx.Index,
+							"err", err.Error(),
+						)
+						continue IterateChains
+					}
+					if !gasPriceIncrease.IsNil() && !gasPriceIncrease.IsZero() {
+						ctx.Logger().Info("GasStabilityPool: updated gas price for pending cctx",
```

Comment:
This might cause a lot of log lines, a max of 200 lines every 100 blocks.
Maybe we can aggregate all this data into a struct and emit that in an event?

Reply:
[7c2084b](https://github.com/zeta-chain/node/pull/1687/commits/7c2084bb0dc1c2aebe605ee6e49439c976190ffe)